### DISCUSSION TITLE
Use Cucumber React Component to render Reqnroll output as HTML

### DISCUSSION THREAD

**Discussion by clrudolphi** [reqnroll/Reqnroll#302] [Ideas]
The implementation of Cucumber Messages support for Reqnroll is making progress. 
I have a fork of Cucumber/HtmlFormatter ported over to .NET and it successfully consumes Messages produced by Reqnroll to produce a (somewhat minimal) HTML report for a feature. The Cucumber/HtmlFormatter uses a React component to render the output of a feature run.

An example of such a report is attached 
![image](https://github.com/user-attachments/assets/6734aea8-3457-4ff8-a6d4-be5cf15ed219)

The HtmlFormatter is a library assembly. I have built a rudimentary console app that consumes an .ndjson file (from Reqnroll) and outputs an HTML file. Code is [dotnet/htmlformatter](https://github.com/clrudolphi/cucumber-html-formatter/tree/add_dotnet/dotnet/htmlformatter)

I have not yet submitted this fork to the Cucumber repository.

I propose that the HtmlFormatter library assembly be submitted to the Cucumber/HtmlFormatter repo (alongside the existing Java, Javascript, and Ruby implementations). Reqnroll would then consume it as a Nuget package.
The console sample is not likely worth contributing to either Cucumber or to Reqnroll. I'm happy to keep it on my own github as a sample or to contribute it to the ExploratoryTestProjects repo within Reqnroll.

An alternative packaging would be to keep the library assembly internal within Reqnroll as we further build out our reporting/integration capabilities.

Thoughts?

**Comment by tazambuja** [comment]
Excellent proposal @clrudolphi  . And very necessary, especially with the end of Specflow support. Have you thought about having an extension in Azure and Jira to publish this report after running tests in the pipeline, like Specflow LivingDoc does/did?

**Comment by MrSoerenAndersen** [comment]
Great to hear that your implementation of Cucumber Messages in ReqnRoll is progressing well. I'm looking forward to migrate our SpecFlow projects to ReqnRoll, including support for a LivingDoc-style report.

Could you share what the final solution in ReqnRoll will look like? For example, will it generate a TestExecution.ndjson file out of the box, or will Cucumber Messages be made available for consumption via plugins or hooks?

At the moment, I'm working on an alternative version of a ReqnRoll LivingDoc report, but I’ll need to finalize integration based on your upcoming implementation.

Can you share any detailed information or examples that could help support this alternative solution?

![image](https://github.com/user-attachments/assets/e16f321f-2b87-44c8-b2a0-bc349ddcd600)

Demo Report: https://expressium.dev/reqnroll/LivingDoc.html
GitHub Project: https://github.com/ExpressiumOSS/ExpressiumLivingDocTestReport

**Reply by gasparnagy** [reply]
The current plan is that the cucumber messages ndjson file can be generated out of the box, no extra plugin is required for that, however you need to set a config file setting to enable this (and you can also specify the file name and location). 

Most probably we will not generate HTML out of the box (but this is not decided yet), but have a command line tool, like the living doc generator was, that picks the ndjson file and can generate a single HTML file out of that.

Does this answer your question? Do you need more details?

**Reply by ExpressiumOSS** [reply]
Yes, absolutely — it sounds great to have an NDJSON file available out of the box. I assume I’ll be able to use my own custom converter or HTML generator to create a LivingDoc report from it.

Do you happen to have any implementation or a complete ReqnRoll NDJSON file that I could take a look at to get started?

**Reply by clrudolphi** [reply]
For examples there are two options you may wish to examine. 
First, feel free to pull down and use the feature-CucumberMessages branch. There are tests that provide proof of compatibility with the Messages spec. Executing those tests will result in ndjson files being written to disk.

2nd: Those tests are based upon cucumber/compatibility-kit which provides the official implementation of test examples for Messages.

The tests within the Cucumber Messages branch in Reqnroll are based upon the compatibility-kit but with some adjustments as Reqnroll doesn't support a few of the features of cucumber.

**Comment by MrSoerenAndersen** [comment]
Hey ReqnRoll - What is the status on the Cucumber Messages ndjson implementation...?

**Reply by gasparnagy** [reply]
In progress. @clrudolphi works on splitting the big PR into smaller ones so that we can review and merge them one by one.

**Reply by clrudolphi** [reply]
The extraction of common/foundational changes is complete. The team is reviewing the main functions of the Messages implementation. 

Regarding your question from earlier: 
>will it generate a TestExecution.ndjson file out of the box, or will Cucumber Messages be made available for consumption via plugins or hooks?

Yes to both questions. We have adopted an approach similar to Cucumber in that we will have 'formatter' plugins. One will write the messages out as ndjson files. Another will emit static HTML files.
